sklearn/datasets/kddcup99.py

```diff
@@ -0,0 +1,414 @@
+        To return the corresponding classical subsets of kddcup 99
+        If None, return the entire kddcup 99 dataset.
+
+    random_state : int, RandomState instance or None, optional (default=None)
+        Random state for shuffling the dataset.
+        If int, random_state is the seed used by the random number generator;
+        If RandomState instance, random_state is the random number generator;
+        If None, the random number generator is the RandomState instance used
+        by `np.random`.
+
+    shuffle : bool, default=False
+        Whether to shuffle dataset.
+
+
+    Returns
+    -------
+    data : Bunch
+        Dictionary-like object, the interesting attributes are:
+        'data', the data to learn and 'target', the regression target for each
+        sample.
+    """
+    kddcup99 = fetch_kddcup99(shuffle=True)
+
+    data = np.c_[kddcup99.data, kddcup99.target]  # XXX to be removed
+
+    if subset == 'SA':
+        s = data[:, -1] == 'normal.'
+        t = np.array([not a for a in s])
+        normal_samples = data[s, :]
+        abnormal_samples = data[t, :]
+
+        n_samples_abnormal = abnormal_samples.shape[0]
+        # selected abnormal samples:
+        r = np.random.randint(0, n_samples_abnormal, 3377)
+        abnormal_samples = abnormal_samples[r]
+        data = np.r_[normal_samples, abnormal_samples]
+        data = np.random.permutation(data)
+
+    if subset == 'SF' or subset == 'http' or subset == 'smtp':
+        # select all samples with positive logged_in attribute:
+        s = data[:, 11] == 1
+        data = np.c_[data[s, :11], data[s, 12:]]
+
+        a0 = np.array([data[:, 0]]).T
+        a2 = np.array([data[:, 2]]).T
+        a4 = np.array([data[:, 4]]).T
+        a5 = np.array([data[:, 5]]).T
+        a6 = np.array([data[:, -1]]).T
+            
+        a0 = np.log(np.asarray(a0 + 0.1, dtype=float))
+        a4 = np.log(np.asarray(a4 + 0.1, dtype=float))
+        a5 = np.log(np.asarray(a5 + 0.1, dtype=float))
+
+        if subset == 'http':
+            s = data[:, 2] == 'http'
+            data = data[s]  # np.c_[data[s, :2], data[s, 3:]]
+            data = np.concatenate((a0, a4, a5, a6), axis=1)
+
+
+        if subset == 'smtp':
+            s = data[:, 2] == 'smtp'
+            data = data[s]  # np.c_[data[s, :2], data[s, 3:]]
+            data = np.concatenate((a0, a4, a5, a6), axis=1)
+
+        if subset == 'SF':
+            data = np.concatenate((a0, a2, a4, a5, a6), axis=1)
+
+    target = data[:, -1]
+    data = data[:, :-1]
+
+    if shuffle:
+        rng = check_random_state(random_state)
+        indices = np.arange(target.shape[0])
+        rng.shuffle(indices)
+        data = data[indices]
+        target = target[indices]
+
+    return Bunch(data=data, target=target)
+
+
+def fetch_kddcup99(subset=None, data_home=None, download_if_missing=True,
+                   random_state=None, shuffle=False):
+
+    """Load the kddcup99 dataset, downloading it if necessary.
```

Comment:
"load" vectorizes and transformes the data. But we can incorporate it into "fetch"...